File: go/sns/ShowSubscriptions/ShowSubscriptions.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
// snippet-start:[sns.go.show_subscriptions]
package main

// snippet-start:[sns.go.show_subscriptions.imports]
import (
    "fmt"

    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/sns"
    "github.com/aws/aws-sdk-go/service/sns/snsiface"
)
// snippet-end:[sns.go.show_subscriptions.imports]

// GetSubscriptions retrieves a list of your Amazon SNS subscriptions
// Inputs:
//     svc is an Amazon SNS service client
// Output:
//     If success, information about your subscriptions and nil
//     Otherwise, nil and an error from the call to ListSubscriptions
func GetSubscriptions(svc snsiface.SNSAPI) (*sns.ListSubscriptionsOutput, error) {
    // snippet-start:[sns.go.show_subscriptions.call]
    result, err := svc.ListSubscriptions(nil)
    // snippet-end:[sns.go.show_subscriptions.call]

    return result, err
}

func main() {
    // snippet-start:[sns.go.show_subscriptions.session]
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))

    svc := sns.New(sess)
    // snippet-end:[sns.go.show_subscriptions.session]

    result, err := GetSubscriptions(svc)
    if err != nil {
        fmt.Println("Got an error retrieving the subscriptions:")
        fmt.Println(err)
        return
    }

    // snippet-start:[sns.go.show_subscriptions.display]
    fmt.Println("Topic ARN")
    fmt.Println("Subscription ARN")
    fmt.Println("-------------------------")
    for _, s := range result.Subscriptions {
        fmt.Println(*s.TopicArn)
        fmt.Println(*s.SubscriptionArn)

        fmt.Println("")
    }
    // snippet-end:[sns.go.show_subscriptions.display]
}
// snippet-end:[sns.go.show_subscriptions]

File: ruby/example_code/sns/show_subscriptions.rb
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
require 'aws-sdk-sns'
require 'logger'

# snippet-start:[sns.Ruby.showSubscription]
# This class demonstrates how to list subscriptions to an Amazon Simple Notification Service (SNS) topic
class SnsSubscriptionLister
  def initialize(sns_client)
    @sns_client = sns_client
    @logger = Logger.new($stdout)
  end

  # Lists subscriptions for a given SNS topic
  # @param topic_arn [String] The ARN of the SNS topic
  # @return [Types::ListSubscriptionsResponse] subscriptions: The response object
  def list_subscriptions(topic_arn)
    @logger.info("Listing subscriptions for topic: #{topic_arn}")
    subscriptions = @sns_client.list_subscriptions_by_topic(topic_arn: topic_arn)
    subscriptions.subscriptions.each do |subscription|
      @logger.info("Subscription endpoint: #{subscription.endpoint}")
    end
    subscriptions
  rescue Aws::SNS::Errors::ServiceError => e
    @logger.error("Error listing subscriptions: #{e.message}")
    raise
  end
end

# Example usage:
if $PROGRAM_NAME == __FILE__
  sns_client = Aws::SNS::Client.new
  topic_arn = 'SNS_TOPIC_ARN' # Replace with your SNS topic ARN
  lister = SnsSubscriptionLister.new(sns_client)

  begin
    lister.list_subscriptions(topic_arn)
  rescue StandardError => e
    puts "Failed to list subscriptions: #{e.message}"
    exit 1
  end
end
# snippet-end:[sns.Ruby.showSubscription]
